### DISCUSSION TITLE
Common Node API

### DISCUSSION THREAD

**Discussion by ansermino** [filecoin-project/FIPs#1032] [Enhancements - Organizational]
This is for discussion around the FIP we are proposing to establish a common API for node implementations. 

A draft PR has been created: #1027 

The current abstract is:
> Lacking a common API for all node implementations, client diversity is hindered as applications are unlikely to be interoperable with different node implementations. Lotus is the defacto reference implementation, but lacks a sufficiently specified interface for other implementers to replicate.
>
> In this FIP a subset of existing RPC methods are formally specified to enable consistency across implementations, and provide clarity on the methods users can expect when interacting with a node.

This effort is partly inspired by our challenges implementing the current Lotus API in Forest. Most notably, the sheer number of methods, a lack of documentation, an incomplete OpenRPC document, and inconsistencies in how types and errors are handled across methods make replicating it especially difficult.

We have chosen to take a path of least resistance (ie. no breaking changes) and directly include a subset of the current interface of Lotus, choosing to omit error handling or access control to reduce complexit. We intend to continue to refine the interface beyond this FIP, with more significant changes in a future `V2` via an additional FIP.

One discussion point is whether this should be an FIP or FRC. Any guidance or commentary here would be welcomed.

**Comment by rvagg** [comment]
> One discussion point is whether this should be an FIP or FRC.

FRC's continue to be under-documented and under-specified and my take on why they exist is that there's a desire for a smoother pathway to publishing something in a central location that may encourage collaboration, coordination and compatibility _without_ requiring the whole ecosystem to agree that it should be adopted.

The new proposed FIP-0001 text has a few more words regarding FRCs that are probably helpful: https://github.com/filecoin-project/FIPs/pull/850/files

IMO this is an FRC effort, we don't need to achieve consensus, what we do in the RPC implementations is not going to break network consensus per se, although it is fairly critical to builders on-top of RPCs, so coordinating in public and seeking input from consumers is good; it just doesn't need everyone to agree and the implementers are in the best position to decide both how much work they want to sign up for as a result of this and the practicalities of the APIs themselves. We should also prefer (I think) to not block on the heavy-weight FIP process to get this in. Let's get a basic agreement, land it, then iterate if we discover problems. Ideally the next step would be to do a v2 API together that's much more focused and consistent.

**Comment by BigLep** [comment]
For visibility, there was a recent FIL Dev Summit talk on this topic: https://www.youtube.com/watch?v=R_sFU97iMzU

**Comment by BigLep** [comment]
I think it's great that FRC / spec for this is being developed.  We'll also want to have a way to test/show conformance.  I'm dropping some notes here from the IPFS community.

IPFS made a push over the last couple of years to have more implementations.  IPFS has a set of [HTTP Gateway specs](https://specs.ipfs.tech/http-gateways/) and they built a conformance test suite and tooling to support it.  Here are some relevant links:
* https://github.com/ipfs/gateway-conformance
* https://conformance.ipfs.tech/
* https://www.youtube.com/watch?v=PmIf77thO_c

While written in Go, it uses a DSL that is very understandable and usable for anyone I think (look at some of the [tests](https://github.com/ipfs/gateway-conformance/tree/main/tests)).  This was important as this conformance test suite would be authored and reviewed by IPFS implementers, some of which don't know/use Go.t

The creators and maintainers of this work are still around (e.g., [IPDX](https://www.ipdx.co/), [IP Shipyard](https://ipshipyard.com/)).  I expect they would be happy to answer questions if we have any.

(That said, if there are better alternatives - that's great.  I'm just passing on a recent option I'm aware of that has worked well for that community.)

I understand there is a precondition to run a Forest, Lotus, etc. node an "offline mode" serving RPC requests from a snapshot.  I heard Forest has this (great!).  Here's an issue to add this capability to Lotus: https://github.com/filecoin-project/lotus/issues/12226

**Reply by BigLep** [reply]
For visibility, Forest RPC test snapshots are documented in https://docs.forest.chainsafe.io/developers/guides/rpc_test_snapshot/

**Comment by BigLep** [comment]
2025-05-22 update based on FDS-6 where FilOz and ChainSafe connected on this topic.  (Much of this is copy/paste from other places.)

1. We're committed to getting https://github.com/filecoin-project/FIPs/pull/1027 merged ASAP but it's going to be in a new PR. @AlexeyKrasnoperov will do necessary (minor) tidy up of that (stored as https://github.com/filecoin-project/FIPs/tree/master/FRCs/frc-0104.md) and attach an OpenRPC spec doc (to be stored in https://github.com/filecoin-project/FIPs/tree/master/resources/frc-104 for now) and we'll get it landed.
2. It's completely agreed that that we should have human authored specs rather than generating specs from go code like has traditionally been done from Lotus.
3. We do want validation tooling that can hit any RPC endpoint and validate conformance ([like IPFS](https://github.com/filecoin-project/FIPs/discussions/1032#discussioncomment-10035397)) and other communities have). This should be able to be run in Forest, Lotus, etc. CIs. Forest has prior starts at this, and when picked up, it should be built assuming an HTTP interface so that it can be run against other implementations like Lotus. When this gets restarted and we have a set of test fixtures, it will be appropriate to create a new filecoin-project repo.
4. Concerning [Lotus RPC API /v1](https://github.com/filecoin-project/lotus/blob/master/documentation/en/api-v1-unstable-methods.md), which is a superset of the Common Node API, and [/v2](https://github.com/filecoin-project/lotus/blob/master/documentation/en/api-v2-unstable-methods.md), which doesn’t have any of the /v1 Common Node API currently, we’ll copy these outside of Lotus as discussed in https://github.com/filecoin-project/lotus/issues/13114.  
5. We can expand out of the FIPs repo at any point where we find ourselves doing complicated enough things or wanting to make the contribution process more frictionless. But we’ll start with that repo for now if we can get away with landing a single doc and figure out the rest later.  Also, we can also move out if another repo gets setup as part of https://github.com/filecoin-project/lotus/issues/13114

The idea here is to not block the common node API on the grand plans of spec separation.  We are fully aware that there are debts across the spec definitions already. There is no need to block the common node API on it.  We also know that getting the separate specs right and integrated into existing places is non-trivial and we don’t have enough capacity to get it right currently. Duplication of spec and slowly reducing emphasis on “stuff” in Lotus seems like the path of least resistance and most impact delivered.

**Reply by BigLep** [reply]
New Common Node API PR: https://github.com/filecoin-project/FIPs/pull/1158
